### PR TITLE
Fix for #19183 to add support for new PHP MongoDB extension in sessions.

### PR DESCRIPTION
**Pull request by omanizer**
| Q | A |
| --- | --- |
| Branch? | 3.0 |
| Bug fix? | yes |
| New feature? | yes (ish) |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | no |
| Fixed tickets | #19183 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 1]
```diff
@@ -132,19 +136,33 @@ public function gc($maxlifetime)
...
-            $this->options['expiry_field'] => $expiry,
-        );
-
-        $this->getCollection()->update(
-            array($this->options['id_field'] => $sessionId),
-            array('$set' => $fields),
-            array('upsert' => true, 'multiple' => false)
-        );
+        $expiry = $this->createDateTime(time() + (int) ini_get('session.gc_maxlifetime'));
+
```
Comment (Simperfit) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Could you use a 
```php
  $fields = array(
        $this->options['time_field'] => $this->createDateTime(),
       $this->options['expiry_field'] => $expiry,
      );
```

Then add the MongoDbBin part ?

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 2]
```diff
@@ -132,19 +136,33 @@ public function gc($maxlifetime)
...
+        if ($this->mongo instanceof \MongoDB\Client) {
+            $fields = array(
+                $this->options['data_field'] => new \MongoDB\BSON\Binary($data, \MongoDB\BSON\Binary::TYPE_OLD_BINARY),
+                $this->options['time_field'] => $this->createDateTime(),
+                $this->options['expiry_field'] => $expiry,
+            );
+
+            $this->getCollection()->updateOne(
+                array($this->options['id_field'] => $sessionId),
```
Comment (Simperfit) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: I guess you could use a variable here too ?

**Comment by Simperfit** [comment]
Could you please update the test according to this class ?

**Comment by ro0NL** [comment]
This is a duplicate of #18788 right?

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php (hunk 1 above)
Comment (omanizer) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Yes.  I'd suggest having 2 different blocks due to the update method name changing too (update vs updateOne as well as needing the 'multiple' property passed in the options parameter for the first one.  I'll modify this to condense the $fields declaration though.

**Comment by omanizer** [comment]
@ro0NL This is a dup of that PR.  I should have looked harder as I didn't see that one when I created this one.  Happy to move things along on this one though if desired.

@Simperfit I've updated the tests accordingly.

**Comment by omanizer** [comment]
Looks like some builds are not passing.  I hypothesize that it's due to this line:
`@requires extension mongo or mongodb` not actually causing the test to be skipped when it may need to be.

Has anyone done an "or" in a require statement before or does anyone know of a better way to accomplish this or any other thoughts?

**Comment by ro0NL** [comment]
@omanizer sure, just check the feedback from that one, looks like want @fabpot wants to move on with this.

**Comment by fabpot** [comment]
As there is no feedback on #18788, @omanizer can you finish this one instead of closing it? You can also get some inspiration from the other PR if that makes sense for you.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php [hunk 3]
```diff
@@ -132,19 +136,30 @@ public function gc($maxlifetime)
...
 
-        $this->getCollection()->update(
-            array($this->options['id_field'] => $sessionId),
-            array('$set' => $fields),
-            array('upsert' => true, 'multiple' => false)
-        );
+        if ($this->mongo instanceof \MongoDB\Client) {
+            $fields[$this->options['data_field']] = new \MongoDB\BSON\Binary($data, \MongoDB\BSON\Binary::TYPE_OLD_BINARY);
+
+            $this->getCollection()->updateOne(
```
Comment (Simperfit) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: You can do the same as before 

```php
$methodName = ($this->mongo instanceof \MongoDB\Client) ? 'updateOne' : 'update';
```

WDYT?

Comment (omanizer) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/MongoDbSessionHandler.php]: Makes sense.  See the latest commit.

**Comment by fabpot** [comment]
Thank you @omanizer.

**Pull request closed without merging**